File: LC/[1340] Jump Game V - DFS.cpp
class Solution {
public:

    int mem[1001];//default to all zero

    int maxJumps(vector<int>& arr, int d) {
        int ret = 0;
        for(int ii=0; ii<arr.size(); ++ii){//try each starting point
            dfs(ii, arr, d);
            ret = max(ret, mem[ii]);
        }
        return ret;
    }

    int dfs(int idx, vector<int> & arr, int D){
        //visited?
        if(mem[idx] != 0) return mem[idx];

        //going right
        int ret = 1;
        for(int kk=1; kk<=D; ++kk){ //for each kk between (jj, ii)
            if(idx+kk >= arr.size()) break;//no need to try further
            if(arr[idx+kk] >= arr[idx]) break;
            ret = max(dfs(idx+kk, arr, D)+1, ret);
        }

        //going left
        for(int kk=1; kk<=D; ++kk){ //for each kk between (jj, ii)
            if(idx-kk < 0) break;//no need to try further
            if(arr[idx-kk] >= arr[idx]) break;
            ret = max(dfs(idx-kk, arr, D)+1, ret);
        }

        mem[idx] = ret;
        return ret;
    }
};

File: LC/[1340] Jump Game V - DFS.java
class Solution {
    private int dfs(int idx, Map<Integer, Integer> idx2cntMemo, int D, int[] arr){
        int N=arr.length;
        if(idx2cntMemo.containsKey(idx)) return idx2cntMemo.get(idx);

        int res=1;//default is itself.
        //jumping left
        for(int jj=idx-1; jj>=Math.max(0, idx-D); --jj){
            if(arr[jj] >= arr[idx]) break;
            res=Math.max(res, 1+dfs(jj, idx2cntMemo, D, arr));
        }

        //jumping right
        for(int jj=idx+1; jj<Math.min(N, idx+D+1); ++jj){
            if(arr[jj] >= arr[idx]) break;
            res=Math.max(res, 1+dfs(jj, idx2cntMemo, D, arr));
        }

        idx2cntMemo.put(idx, res);
        return res;

    }
    public int maxJumps(int[] arr, int d) {
        int N=arr.length;

        Map<Integer, Integer> idx2cntMemo = new HashMap<>();

        //we need to try each index 
        int res=0;
        for(int ii=0; ii<N; ++ii){
            res=Math.max(res, dfs(ii, idx2cntMemo, d, arr));
        }

        return res;
    }
}